github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/20084359?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/20084359?buildTab=artifacts#meta) on refs/heads/master @ [a2947b4a9383](https://github.com/cockroachdb/pebble/commits/a2947b4a9383):


```
          key_format=testkeys
          strictfs=true
          ingest_using_apply=true
          delete_sized=true
          replace_single_delete=true
          threads=7
          disable_block_property_collector=true
          enable_value_blocks=true
          async_apply_to_db=true
          seed_efos=459089948038828765
          ingest_split=true
          use_excise=true
        
        ===== OPS =====
        /artifacts/meta/250702-044254.780496354451/ops
        ===== HISTORY =====
        /artifacts/meta/250702-044254.780496354451/random-028/history
        ...[2672 lines skipped]...
        iter14.InternalNext() // <nil> #2025
        db1.Get("grfwcwpcwi@1") // [""] pebble: not found #2027
        db1.RangeKeySet("jazzdooiqnj", "luundthe", "@3", "ikolar") // <nil> #2028
        iter14.Last() // [false] <nil> #2029
        iter14.SeekLT("ljoyrm@5", "hwvdoqmzqt@5") // [invalid] <nil> #2030
        db1.SingleDelete("xgczmuc", false /* maybeReplaceDelete */) // <nil> #2031
        iter14.Prev("") // [false] <nil> #2032
        iter14.Prev("") // [false] <nil> #2033
        iter14.SeekGE("bqahzagpn", "") // [false] <nil> #2035
        iter14.SetBounds("brmlis", "brmlis") // <nil> #2037
        iter14.SeekLT("brmlis", "") // [false] <nil> #2038
        db1.LogData("txyuybycpch") // <nil> #2039
        iter14.First() // [false] <nil> #2041
        iter14.SeekPrefixGE("jpzwgjzq") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2042
        iter14.NextPrefix() // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2043
        iter14.Next("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2044
        db1.Get("hzelmrkb@1") // [""] pebble: not found #2049
        iter14.SeekPrefixGE("jwrtrdchqdvh@4") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2050
        db1.RatchetFormatMajorVersion(024) // <nil> #2052
        iter14.SeekPrefixGE("eyvfjleyl@7") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2054
        db1.SingleDelete("bprjstyldjo", false /* maybeReplaceDelete */) // <nil> #2056
        iter14.Last() // [false] <nil> #2057
        // 04:42:55.197 INFO: [JOB 75] blob file (B000066, 000170) rewrite error: blob file B000066 became unreferenced: pebble: compaction cancelled by a concurrent operation, will retry compaction
        // 04:42:55.197 FATAL: obsoleteBlobs is not sorted
        // 04:42:55.197 INFO: [JOB 75] sstable deleted 000173
        // 04:42:55.197 INFO: [JOB 75] blob file deleted 000174
        // 04:42:55.197 INFO: [JOB 75] blob file deleted 000171
        iter14.SetOptions("brmlis", "irtigl@1", 0 /* key types */, "", "", true /* use L6 filters */, "" /* masking suffix */) // <nil> #2058
        iter15.SeekGE("eobd@9", "") // [false] <nil> #2180
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/250702-044254.780496354451/random-028 --try-to-reduce -v
--- FAIL: TestMeta/execution/random-028 (0.42s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1751431374780748741 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-532

--- a/blob_rewrite.go
+++ b/blob_rewrite.go
@@ -215,16 +215,18 @@ func (c *blobFileRewriteCompaction) Execute(jobID JobID, d *DB) error {
 	// Ensure we clean up the blob file we created on failure.
 	if err != nil {
 		if objMeta.DiskFileNum != 0 {
-			d.mu.versions.obsoleteBlobs = append(d.mu.versions.obsoleteBlobs, obsoleteFile{
-				fileType: base.FileTypeBlob,
-				fs:       d.opts.FS,
-				path:     d.objProvider.Path(objMeta),
-				fileNum:  objMeta.DiskFileNum,
-				// We don't know the size of the output blob file--it may have
-				// been half-written. We use the input blob file size as an
-				// approximation for deletion pacing.
-				fileSize: c.input.Physical.Size,
-				isLocal:  true,
+			d.mu.versions.obsoleteBlobs = mergeObsoleteFiles(d.mu.versions.obsoleteBlobs, []obsoleteFile{
+				{
+					fileType: base.FileTypeBlob,
+					fs:       d.opts.FS,
+					path:     d.objProvider.Path(objMeta),
+					fileNum:  objMeta.DiskFileNum,
+					// We don't know the size of the output blob file--it may have
+					// been half-written. We use the input blob file size as an
+					// approximation for deletion pacing.
+					fileSize: c.input.Physical.Size,
+					isLocal:  true,
+				},
 			})
 		}
 	}

